---
title: "Fungal OTU's across life stages of boreal toad: published and supplemental analyses"
author: "Alexandra Alexiev"
date: "12/17/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = TRUE, # evaluate code chunks
                      include = TRUE, # include the console output of the code in the final document
                      echo = TRUE, # include the code that generated the report in the final report
                      warning = FALSE, # include warnings
                      message = FALSE, # include console messages
                      collapse = TRUE, # Merge code blocks and output blocks, if possible.
                      dpi = 300, # the default figure resolution
                      fig.dim = c(9, 5), # the default figure dimensions
                      out.width = '98%', # the default figure output width
                      out.height = '98%', # the default figure output height
                      cache = TRUE) # save the calculations so that kniting is faster each time. (Sometimes this option can cause issues and images won't reflect the most recent code changes, if this happens, just delete the *_cache folder and reknit the code.)
```

```{r}
#packages and set working directory
library(vegan)
library(ggplot2)
library(mctoolsr)
library(tidyr)
library(tibble)
library(gridExtra)
library(dplyr)
library(ggpubr)
library(cowplot)
library(kableExtra)
library(RColorBrewer)
library(BiodiversityR)
set.seed(33333)
work_dir <- "/Users/alal3493/Documents/Projects/DevelopmentalBorealToad/FungalDevData/02_PublishedAnalysis/"
seq_dir <- "/Users/alal3493/Documents/Projects/DevelopmentalBorealToad/FungalDevData/00_SeqProcessing/"
```


## Basic measures of highly abundant or present taxa

```{r}
#load in starting files
mapfp_all <- paste0(work_dir,"Allsamps_clean_maptab.txt")
tabfp_all <- paste0(work_dir,"Allsamps_clean_otutab.txt")
input_all <- load_taxa_table(tab_fp = tabfp_all, map_fp = mapfp_all) # this file has both toad and environment sample types in it (172 samples)
mapfp_t <- paste0(work_dir,"Toad_clean_maptab.txt")
tabfp_t <- paste0(work_dir,"Toad_clean_otutab.txt")
input_toad <- load_taxa_table(tab_fp = tabfp_t, map_fp = mapfp_t) # this file only has toad samples in it (124 samples)

# order by life stage
life_list <- c("Soil", "Sediment", "Water",
               "Eggs.11.15", "Tadpole.20.22",
               "Tadpole.23.25", "Tadpole.25.27",
               "Tadpole.27.29", "Tadpole.29.31",
               "Tadpole.31.35", "Tadpole.36.39",
               "Metamorph.40.46", "Subadult", "Adult")
input_all$map_loaded$Gosner_Stage <-
    factor(input_all$map_loaded$Gosner_Stage,
           levels=life_list)
input_toad$map_loaded$Gosner_Stage <-
    factor(input_toad$map_loaded$Gosner_Stage,
           levels=life_list)
# and order the other column I'll be using as well, where tadpoles are lumped
samptype_list <- c("Soil", "Sediment", "Water", "Eggs",
                   "Tadpole", "Metamorph", "Subadult", "Adult")
input_all$map_loaded$Life_Stage_Simplified <-
    factor(input_all$map_loaded$Life_Stage_Simplified,
           levels=samptype_list)
input_toad$map_loaded$Life_Stage_Simplified <-
    factor(input_toad$map_loaded$Life_Stage_Simplified,
           levels=samptype_list)
```

```{r}
# number of total fungal taxa found in all samples
nrow(input_all$data_loaded) # 2243
nrow(input_toad$data_loaded) # 1703

# top ten taxa on toads and in environment
toptaxa_all <- return_top_taxa(input = input_all, number_taxa = 10)
toptaxa_toad <- return_top_taxa(input = input_toad, number_taxa = 10)
kable(toptaxa_all, caption="Top ten taxa in toad and environment samples") %>% 
    kable_styling(bootstrap_options = c("striped","bordered"), 
                  full_width = F, position = "left")
kable(toptaxa_toad, caption="Top ten taxa in toad samples") %>% 
    kable_styling(bootstrap_options = c("striped","bordered"), 
                  full_width = F, position = "left")
```


# Q1: What are the factors associated with variation in fungal communities?

## Alpha diversity differences between life stages and environment samples

Below is a supplementary figure comparing three alpha diversity measures across all the sample types. Since site is a confounding factor here, given exploratory analyses, I analyzed with blocking by site ("strata").

```{r}
# calculate alpha diversity measures
input_all$map_loaded$rich = calc_diversity(
  tax_table = input_all$data_loaded, metric = 'richness')
input_all$map_loaded$simpson = calc_diversity(
  tax_table = input_all$data_loaded, metric = 'simpson')
input_all$map_loaded$shannon = calc_diversity(
  tax_table = input_all$data_loaded, metric = 'shannon')

# Kruskall-Wallis test on life stages, strata-site
# species richness
krusk_rich <- kruskal.test(input_all$map_loaded$rich ~ 
                              input_all$map_loaded$Gosner_Stage)
bonpval_rich <- p.adjust(p = krusk_rich$p.value, method = "bonferroni", 
         n = length(unique(input_all$map_loaded$Gosner_Stage)))
krusk_rich; bonpval_rich

# simpson (evenness)
krusk_simp <- kruskal.test(input_all$map_loaded$simpson ~ 
                              input_all$map_loaded$Gosner_Stage)
bonpval_simp <- p.adjust(p = krusk_simp$p.value, method = "bonferroni", 
         n = length(unique(input_all$map_loaded$Gosner_Stage)))
krusk_simp; bonpval_simp

# shannon (both richness and evenness)
krusk_shan <- kruskal.test(input_all$map_loaded$shannon ~ 
                              input_all$map_loaded$Gosner_Stage)
bonpval_shan <- p.adjust(p = krusk_shan$p.value, method = "bonferroni", 
         n = length(unique(input_all$map_loaded$Gosner_Stage)))
krusk_shan; bonpval_shan

# GRAPHING
# Species richness graphs: mean number of OTUs found per sample type
liferich <- ggplot(input_all$map_loaded, aes(Gosner_Stage, rich)) + 
  geom_boxplot() + xlab("") + 
  ylab("Fungal OTU Richness") + ggtitle("A") +
  geom_text(x=7,y=194, size=3,
            label=paste0("KW chi-sq stat=",signif(krusk_rich$statistic, digits=5),
                         "\n","p=",signif(bonpval_rich, digits=3))) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, colour = "black"),
        axis.text.y = element_text(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"))

# Simpson's: measures species/OTU evenness, or how close in number each species in the environment
lifesimp <- ggplot(input_all$map_loaded, aes(Gosner_Stage, simpson)) + 
  geom_boxplot() + xlab("Sample Type") + 
  ylab("Simpson's Diversity Metric") +  ylim(0,1.2) + ggtitle("B") +
  geom_text(x=7,y=1.15, size=3,
            label=paste0("KW chi-sq stat=",signif(krusk_simp$statistic, digits=5),
                         "\n","p=",signif(bonpval_simp, digits=3))) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, colour = "black"),
        axis.text.y = element_text(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"))

# Shannon: combination of richness and evenness
lifeshan <- ggplot(input_all$map_loaded, aes(Gosner_Stage, shannon)) + 
  geom_boxplot() + xlab("") + 
  ylab("Shannon Diversity Index") + ggtitle("C") +
  geom_text(x=7,y=4.1, size=3,
            label=paste0("KW chi-sq stat=",signif(krusk_shan$statistic, digits=5),
                         "\n","p=",signif(bonpval_shan, digits=3))) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, colour = "black"),
        axis.text.y = element_text(colour = "black"),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"))

alpha_panel <- grid.arrange(liferich, lifesimp, lifeshan, ncol = 3)
alpha_panel
# ggsave(paste0(work_dir,"Figs_Tables/AllAlphaDiv.png"), alpha_panel, dpi=300,
#        width=10, height=7)
```

# Pairwise OTU richness Stats

- Here, I'm using a two-sample Mann-Whitney test (or unpaired wilcoxon test). This is a non-parametric test of whether two populations are different across two different life stages, while still assuming those populations are independent. Since my samples came from different animals at different life stages, I need to make this assumption, so I cannot use a paired wilcoxon test.
- I calculated Bonferroni (signif if p<=0.05) and FDR (p<=0.05) p-value corrections, used both together to figure out how big of a difference there was between each life stage.

```{r}
# Are pairwise alpha diversities same or different across life stages?
nlevels(input_all$map_loaded$Gosner_Stage) # 14
cats <- levels(input_all$map_loaded$Gosner_Stage) # define category levels
cats_pairs <- combn(cats, 2) # make pairs of these
pvals <- c() # output df for pvalues
Wstat <- c() # output df for Wstat
for ( i in 1:ncol(cats_pairs) ) { # for each pair of sample types
    pair <- cats_pairs[, i] # define the pair
    Gosner_rich <- input_all$map_loaded %>% # take the mapping file
    dplyr::select(Gosner_Stage, rich) %>% # select these two columns from it
    filter(Gosner_Stage %in% pair) # filter the sample types in the pair
    wil <- wilcox.test(rich ~ Gosner_Stage, data = Gosner_rich, paired = F, exact = F) # run Wilcoxon test
    pvals <- c(pvals, wil$p.value) # save the p val
    Wstat <- c(Wstat, wil$statistic) # save the W statistic
}
results <- data.frame(t(cats_pairs), Wstat, pvals) # make a dataframe of the useful outputs
results$pvalBon = pvals * length(pvals) # calc bonferroni pval correction
results$pvalFDR = round(pvals * (length(pvals) / rank(pvals, ties.method = "average")),
                          3) # calc FDR pval correction
results <- results %>%
  arrange(pvalBon, pvalFDR) %>% # arrange in ascending order by the corrected pvals
  filter(pvalBon <= 0.05) %>%
  filter(pvalFDR <= 0.05) %>% # filter by the two corrected pvals being <= 0.05
  dplyr::select(-pvals) # remove pval column
# write.table(results, paste0(work_dir, "Figs_Tables/pairedMW_alphadiv.txt"), 
#             sep="\t", quote = F, row.names = F)
```

## Make three paneled figure with alpha and beta diversity measures
```{r}
# make color palette
nlevels(input_all$map_loaded$Life_Stage_Simplified)
brewer.pal(n = 8, name = "Dark2") # display hexcodes for the pallete I'm using but want to change slightly
samptype_cols <- c("#666666", "#A6761D", 
                   "#3a6b94", "#66A61E", 
                   "#E7298A", "#7570B3", 
                   "#D95F02", "#1B9E77")

# make OTU richness figure
OTUrich <- ggplot(input_all$map_loaded, aes(Gosner_Stage, rich)) + 
  geom_boxplot(alpha=0.6, outlier.shape=NA, aes(fill=Life_Stage_Simplified)) + xlab("") + 
  ylab("Fungal OTU Richness") + ggtitle("A") +
  geom_point(aes(fill=Life_Stage_Simplified), size=3, shape=21, position=position_jitter(
    width=0.25
  )) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, colour = "black"),
        axis.text.y = element_text(colour = "black"),
        axis.title = element_text(size=16),
        axis.text = element_text(size=14),
        legend.title = element_text(size=16),
        legend.text = element_text(size=14),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"),
        plot.title = element_text(size = 30),
        legend.position="none") +
  labs(fill="Sample Type") +
  scale_fill_manual(values=samptype_cols) +
  scale_x_discrete(labels=c("Soil", "Sediment", "Water",
                            "Eggs", "Tadpole (20-22)",
                            "Tadpole (23-25)", "Tadpole (25-27)",
                            "Tadpole (27-29)", "Tadpole (29-31)",
                            "Tadpole (31-35)", "Tadpole (36-39)",
                            "Metamorph (40-46)", "Subadult", "Adult"))

# make NMDS plot
# create distance matrix
md_transformed <- t(sqrt(input_all$data_loaded))
dm <- vegdist(md_transformed, method = "bray")

# run NMDS
md.nmds <- metaMDS(dm, k = 3, trymax = 1000) #solution reached, stress=0.19

#prepare NMDS points and values for graphing
md.nmds.points <- md.nmds$points %>% # take the NMDS points
  data.frame() %>% # make a data frame
  rownames_to_column(var = "mysamples") %>%
  mutate(SampleID = mysamples) %>%
  column_to_rownames(var = "mysamples")
meta <- input_all$map_loaded %>%
  rownames_to_column("SampleID")
md.nmds.metadata <- inner_join(x=md.nmds.points, y=meta, by = "SampleID") %>%
  group_by(Life_Stage_Simplified) %>%
  ungroup() %>%
  dplyr::select(SampleID, MDS1, MDS2, everything()) # select these columns only
md.nmds.metadata.unq <- md.nmds.metadata %>%
  dplyr::select(Life_Stage_Simplified, Type) %>%
  unique()
md.nmds.metadata.unq$Life_Stage_Simplified <- 
  factor(md.nmds.metadata.unq$Life_Stage_Simplified, 
                          levels=samptype_list)

# get stress value
stress.md = paste("stress =", round(md.nmds$stress, digits = 4))

# get NMDS hulls
group.chulls <- plyr::ddply(md.nmds.metadata, "Life_Stage_Simplified", 
                            function(df) df[chull(df$MDS1, df$MDS2), ])

# Plot NMDS
nmds.plot <- ggplot(md.nmds.metadata, aes(x = MDS1, y = MDS2, shape = Type,
                                          colour = Life_Stage_Simplified)) +
  geom_polygon(data = group.chulls, 
               aes(fill=Life_Stage_Simplified), 
               alpha = 0.15, linetype=0) +
  geom_point(size=3) +
  annotate("text", x = Inf, y = Inf, label = stress.md, hjust = 1, vjust = 1) +
  ggtitle("C") + 
  labs(colour="Sample Type") +
  scale_colour_manual(values=samptype_cols) +
  scale_fill_manual(values=samptype_cols) +
  guides(fill=F, color = guide_legend(order = 1, override.aes=list(shape=15, size=6)), 
         shape = guide_legend(order = 0)) +
  scale_shape_discrete(name = "", labels = c("Environment", "Toad")) +
  theme(axis.title = element_text(size=16),
        axis.text = element_text(size=14),
        legend.title = element_text(size=16),
        legend.text = element_text(size=14),
        plot.title = element_text(size = 30))


# Make dispersion graph
# PERMDisp on sample type, strata=site
#create B-C species abundance matrix
input_all_rel <- convert_to_relative_abundances(input_all)
BC <- calc_dm(input_all_rel$data_loaded, method = "bray")
life.disper.lump <- betadisper(BC, input_all$map_loaded$Life_Stage_Simplified)
permutest(life.disper.lump)
PermDisp_life_lump <- TukeyHSD(life.disper.lump) #Tukey post-hoc
kable(PermDisp_life_lump$group) %>%
  kable_styling(bootstrap_options = c("striped","bordered"), 
                full_width = F, position = "left")
# write.table(PermDisp_life_lump$group,
#             paste0(work_dir,"Figs_Tables/PermDisp_life_lump.txt"), sep="\t")
Disper1 <- data.frame(life.disper.lump$distances)
colnames(Disper1) <- "dispers"
Disper2 <- Disper1 %>%
  rownames_to_column("SampleID") %>%
  inner_join(y=meta, 
             by="SampleID")

disper_bars <- ggplot(data=Disper2, aes(x=Life_Stage_Simplified, y=dispers)) +
  geom_boxplot(aes(fill=Life_Stage_Simplified), alpha=0.6,outlier.shape=NA) + 
  ylab("Dispersion") + xlab("") + ggtitle("B") +
  geom_point(aes(fill=Life_Stage_Simplified), size=3, shape=21, position=position_jitter(
    width=0.25
  )) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1, colour = "black"),
        axis.text.y = element_text(colour = "black"),
        axis.title = element_text(size=16),
        axis.text = element_text(size=14),
        legend.title = element_text(size=16),
        legend.text = element_text(size=14),
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        panel.background = element_blank(),
        axis.line = element_line(colour = "black"),
        plot.title = element_text(size = 30)) +
  labs(fill="Sample Type") +
  scale_fill_manual(values=samptype_cols)
legend <- arrangeGrob(get_legend(disper_bars))
disper_bars_noleg <- disper_bars + theme(legend.position = "none")


# Put all into paneled graph with annotations and shared legend
diversity_fig <- ggpubr::ggarrange(arrangeGrob(OTUrich, disper_bars_noleg), 
                              legend, nmds.plot, ncol=3, 
               widths=c(4.5,1.5,9))
diversity_fig
ggsave(paste0(work_dir,"Figs_Tables/FungDivFig.png"), diversity_fig, dpi=800,
       width=16, height=7)
```

## Permanova for the above plot

```{r}
# PERMANOVA on life stage, strata=site
Perm <- adonis(BC ~ input_all$map_loaded$Life_Stage_Simplified, 
                        strata = input_all$map_loaded$Location)
kable(Perm$aov.tab) %>%
  kable_styling(bootstrap_options = c("striped","bordered"), 
                full_width = F, position = "left")

# also show the pairwise PERMANOVA
pw_perm <- read.delim(paste0(work_dir,"Figs_Tables/PW_PERM.txt"))
kable(pw_perm) %>%
  kable_styling(bootstrap_options = c("striped","bordered"), 
                full_width = F, position = "left")
```


# Q2: Are there transient versus symbiotic OTU's on boreal toad skin?

Which fungi are more likely to be symbionts in that they are associated with the host as opposed to transiently moving through the host system?

To answer this question, I'm using ANCOM, an indicator species analysis for microbial sequence datasets.
- Uses log-ratio transformation to somewhat account for compositionality of the data, which helps decrease the false discovery rate.
- Uses K-W and wilcoxon tests to determine significance.
- Seems to be more stringent than other methods, like LEfSe or macroecology indicator species analysis, in terms of how many OTU's pass the thresholds.
- Does not need rarefied data, but I am using my rarefied data here.
- ANCOM is running a bunch of pairwise tests and the W stat is how many of those pairwise tests passed the "significance" threshold set in the initial script when I ran it.

## Run ANCOM
```{r}
# Ancom needs Sample.ID to be the header starting at A1, which means changing it in both files and transposing the taxa table.
# export_taxa_table(input = input_all, "taxa_table_ancom.txt", "map_ancom.txt")
# Did changes in excel, and changed OTU ID's to have the taxonomy and OTU # (for future reference)

# re-read in clean files
OTU_ancom <- read.delim(paste0(work_dir,"02_hypothesisbasedanalsysis/otu_ancom.txt"),
                          sep = "\t", header = T)
map_ancom <- read.delim(paste0(work_dir,"02_hypothesisbasedanalsysis/map_ancom.txt"),
                        sep = "\t", header = T)

# source the code, downloaded from: https://github.com/mech3132/MSC_code/blob/master/ANCOM_updated_code_MYC.R
source(paste0(work_dir,"02_hypothesisbasedanalsysis/ANCOM_updated_code_MYC.R"))

#   1) using just sample type as life or env
# run ancom with parameters for no repeated data, no adjustment (so it'll use a Wilcoxon test)
comparison_test_lifeenv <- ANCOM.main.myc(OTUdat = OTU_ancom,
                                      Vardat = map_ancom,
                                      adjusted = F,
                                      repeated = F,
                                      main.var = "Type",
                                      adj.formula = NULL,
                                      repeat.var = NULL,
                                      longitudinal = F,
                                      random.formula = NULL,
                                      multcorr = 2,
                                      sig = 0.05,
                                      prev.cut = 0.90)

# This tells you if your data has a lot of zeroes/skew, as most microbial data do.
# everything below the red line gets removed.
comparison_test_lifeenv$PLot.zeroes
# This tells you which (colored) OTU's have a high effect size and significance
comparison_test_lifeenv$Plot.volcano
comparison_test_lifeenv$Plot.volcano.0.9
# this helps you determine if the critical w was chosen correctly. We want the valley between a bimodal distribution, in theory.
comparison_test_lifeenv$Plot.critical.w 
# I think W=~20 actually would be better, especially with the clumping of the points in the previous plots. I'll take this into account in the write out file.
# write.table(comparison_test_lifeenv$W.taxa, paste0(work_dir,"02_hypothesisbasedanalsysis/indic_taxabunds_lifeenv.txt"), sep="\t", quote=F)

#   2) ANCOM so that each life stage gets compared individually to each environmental sample type. Caveat: this uses a Kruskall-Wallis test so it's not a pairwise comparison (i.e. it won't identify OTU's that are different between e.g. water and tadpole)
comparison_test_lifestages <- ANCOM.main.myc(OTUdat = OTU_ancom,
                                      Vardat = map_ancom,
                                      adjusted = F,
                                      repeated = F,
                                      main.var = "Life_Stage_Simplified",
                                      adj.formula = NULL,
                                      repeat.var = NULL,
                                      longitudinal = F,
                                      random.formula = NULL,
                                      multcorr = 2,
                                      sig = 0.05,
                                      prev.cut = 0.90)
# write.table(comparison_test_lifestages$W.taxa, paste0(work_dir,"02_hypothesisbasedanalsysis/indic_taxabundsnew.txt"), sep="\t", quote=F)
```

## Graph Volcano plot
```{r}
#make a volcano plot showing the associated microbes
otu.names <- comparison_test_lifeenv$W.taxa$otu.names
W_stat <- comparison_test_lifeenv$W.taxa$W_stat
W_f <- comparison_test_lifeenv$W.taxa$W_f
W_frame <- data.frame(otu.names,W_stat,W_f,row.names=NULL)
# in excel, add column for whether they are transient, fac, obl; read in that file
# write.table(W_frame, paste0(work_dir,"Figs_Tables/Volcano_plot/volcano_plot.txt"), sep="\t")
W_frame_V2 <- read.delim(paste0(work_dir,"Figs_Tables/Volcano_plot/volcano_plot_V2.txt"))

26-(26-18)/2 # want to put interept between 18 and 26 because W=26+ is significant and 18 and less is not
# figure out order of ecol for legend
Ecol_list <- c("Obligate", "Facultative", "Transient", "Transient, not significant")
W_frame_V2$Ecol <- factor(W_frame_V2$Ecol, levels=Ecol_list)
# graph volcano plot
volcano.plot <- ggplot(W_frame_V2) +
  geom_point(aes(x=W_f, y=W_stat, color=Ecol)) +
  xlab("CLR F statistic (Effect size)") +
  ylab("W statistic") +
  labs(color = "") +
  scale_color_manual(values=c("#0072B2","#56B4E9", "#CC79A7", "#000000"))+ 
  geom_hline(mapping=aes(yintercept=22), col="red", lty=2)
volcano.plot
# ggsave(paste0(work_dir,"Figs_Tables/Volcano_plot/volcano_plot_color.png"),
#        volcano.plot, dpi=400, width=7, height=5)
```

## Graph relative abundances for significant fungal OTU's
I used excel to clean up the output file "indic_taxabunds_lifeenv.txt" (ANCOM output file) into "indic_taxabunds.txt" which has the information I need moving forward.
```{r}
# read in cleaned up file with relative abundances
indic_abunds <- read.delim(paste0(work_dir,"02_hypothesisbasedanalsysis/indic_taxabunds.txt"), header=T) %>%
  gather(key="Sample.Type", value="Taxon.Relative.Abundance", -c("OTU","Ecol")) %>%
  mutate(Type=case_when(Sample.Type=="Soil" ~ "Environment",
                        Sample.Type=="Sediment" ~ "Environment",
                        Sample.Type=="Water" ~ "Environment", 
                       Sample.Type=="Eggs" ~ "Toad",
                       Sample.Type=="Tadpole" ~ "Toad", 
                       Sample.Type=="Metamorph" ~ "Toad", 
                       Sample.Type=="Subadult" ~ "Toad", 
                       Sample.Type=="Adult" ~ "Toad"))
indic_abunds$Sample.Type <-
    factor(indic_abunds$Sample.Type,
           levels=c("Soil", "Sediment", "Water", "Eggs", "Tadpole",
                    "Metamorph", "Subadult", "Adult"))

# make three sets of faceted graphs and put on a plot together
obl_abunds <- dplyr::filter(indic_abunds, Ecol == "Obligate")
fac_abunds <- dplyr::filter(indic_abunds, Ecol=="Facultative")
tran_abunds <- dplyr::filter(indic_abunds, Ecol=="Transient")
Obl_g <- ggplot(obl_abunds, aes(x=Sample.Type, y=Taxon.Relative.Abundance, 
                           group=OTU, fill=Type)) +
  geom_bar(stat="identity") +
  xlab("") + ylab("") +
  scale_fill_manual(values=c("#548235", "#C55A11"), name="Sample Type") +
  theme(axis.text.x=element_text(angle=90, hjust=1, size=20), 
        axis.title = element_text(size=30),
        axis.text.y=element_text(size=20),
        strip.text = element_text(size = 20),
        legend.position = "none") +
  facet_wrap(~OTU, scales="free_y", ncol=3, labeller = label_wrap_gen(width=15))
Fac_g <- ggplot(fac_abunds, aes(x=Sample.Type, y=Taxon.Relative.Abundance, 
                           group=OTU, fill=Type)) +
    geom_bar(stat="identity") +
    xlab("") + ylab("Taxon Relative Abundance") + 
    scale_fill_manual(values=c("#548235", "#C55A11"), name="Sample Type") +
    theme(axis.text.x=element_text(angle=90, hjust=1, size=20), 
          axis.title = element_text(size=30),
          axis.text.y=element_text(size=20),
          strip.text = element_text(size = 20),
          legend.title = element_text(size=30),
          legend.text = element_text(size=20)) +
    facet_wrap(~OTU, scales="free_y", ncol=3, labeller = label_wrap_gen(width=15))
Tran_g <- ggplot(tran_abunds, aes(x=Sample.Type, y=Taxon.Relative.Abundance, 
                           group=OTU, fill=Type)) +
    geom_bar(stat="identity") +
    xlab("") + ylab("") + 
    scale_fill_manual(values=c("#548235", "#C55A11"), name="Sample Type") +
    theme(axis.text.x=element_text(angle=90, hjust=1, size=20), 
          axis.title = element_text(size=30),
          axis.text.y=element_text(size=20),
          strip.text = element_text(size = 20),
          legend.position = "none") +
    facet_wrap(~OTU, scales="free_y", ncol=3, labeller = label_wrap_gen(width=15))
Obl_g; Fac_g; Tran_g
# pull out the legend (same one for all of the graphs)
legend_abund <- arrangeGrob(get_legend(Fac_g))
# ggsave(paste0(work_dir,"Figs_Tables/Symbionts_fig/legend_abund.png"), legend_abund, dpi=400,
#        width=3, height=3)
Facg_noleg <- Fac_g + theme(legend.position = "none")

#save all these graphs and rearrange how I want them in powerpoint
# ggsave(paste0(work_dir,"Figs_Tables/Symbionts_fig/obl_graph.png"),
#        Obl_g, dpi=400, width=12, height=4.5)
# ggsave(paste0(work_dir,"Figs_Tables/Symbionts_fig/fac_graph.png"),
#        Facg_noleg, dpi=400, width=12, height=7)
# ggsave(paste0(work_dir,"Figs_Tables/Symbionts_fig/tran_graph.png"),
#        Tran_g, dpi=400, width=12, height=4.5)
# Assembled these files into the published graph using powerpoint
```

As an extension of the above graph, make a rank abundance based plot using the data, since ANCOM is actually using rank abundances to identify indicator OTU's
```{r}
# creat a data frame with the abundance values from ANCOM (ANCOM caluclated these)
abund_ENV <- comparison_test_lifeenv$W.taxa$env_relAbund
abund_TOAD <- comparison_test_lifeenv$W.taxa$life_relAbund
otu.names <- comparison_test_lifeenv$W.taxa$otu.names
W_stat <- comparison_test_lifeenv$W.taxa$W_stat
relabunds_ANCOM <- data.frame(otu.names, abund_ENV, abund_TOAD, W_stat)

# calculate rank abundance in Toad samples of each OTU
order.abund_TOAD <- order(relabunds_ANCOM$abund_TOAD,relabunds_ANCOM$otu.names)
relabunds_ANCOM$rank_TOAD <- NA # add as new column for plotting
relabunds_ANCOM$rank_TOAD[order.abund_TOAD] <- 1:nrow(relabunds_ANCOM)

# calculate rank abundance in Environmental samples of each OTU
order.abund_ENV <- order(relabunds_ANCOM$abund_ENV,relabunds_ANCOM$otu.names)
relabunds_ANCOM$rank_ENV <- NA # add as new column for plotting
relabunds_ANCOM$rank_ENV[order.abund_ENV] <- 1:nrow(relabunds_ANCOM)

# filter only significant OTU's (those with W stat greater than 26)
sign_ANCOM_rankabunds <- relabunds_ANCOM %>%
  dplyr::filter(W_stat >= 26)

# plot significant ANCOM data
# rank abundance plot of just the OTU's in toads
rank_toads <- ggplot(sign_ANCOM_rankabunds) + 
  geom_point(aes(x=rank_TOAD, y=abund_TOAD, color=otu.names)) +
  xlab("rank abundance on toad samples") +
  ylab("relative abundance on toad samples")

# rank abundance plot of just the OTU's in env
rank_env <- ggplot(sign_ANCOM_rankabunds) + 
  geom_point(aes(x=rank_ENV, y=abund_ENV, color=otu.names)) +
  xlab("rank abundance on environmental samples") +
  ylab("relative abundance on environmental samples")

# rank abundance of each of toad and env on same graph
rank_plot <- ggplot(sign_ANCOM_rankabunds) + 
  geom_point(aes(x=rank_ENV, y=rank_TOAD, color=otu.names)) +
  xlab("rank abundance on environmental samples") +
  ylab("rank abundance on toad samples")


## this isn't actually showing what I thiiiink it's showing, try OPTION 2
# try it by calculating relative then rank abundance for each OTU in the whole dataset
# relative abundance in toad samples
spe <- t(input_toad$data_loaded)
spe.pres <- apply(spe > 0, 2, sum)
spe.relf <- 100*spe.pres/nrow(spe)

# filter input all for only env samples
input_env <- filter_data(input_all, filter_cat = "Type", 
                      filter_vals = "life") #48 samples remaining
# relative abundance in env samples
spe2 <- t(input_env$data_loaded)
spe.pres2 <- apply(spe2 > 0, 2, sum)
spe.relf2 <- 100*spe.pres2/nrow(spe2)

# clean up data
spe_rel_t <- spe.relf %>%
  as.data.frame() %>%
  rownames_to_column(var = "otu.name")
names(spe_rel_t)[2] <- "rel_abund_toad" # name the second column something
spe_rel_e <- spe.relf2 %>%
  as.data.frame() %>%
  rownames_to_column(var = "otu.name")
names(spe_rel_e)[2] <- "rel_abund_env"
merged_spe <- merge(spe_rel_t, spe_rel_e, by="otu.name")

# calculate rank abundance in Toad samples of each OTU
order.abund_TOAD <- order(merged_spe$rel_abund_toad,as.vector(merged_spe$otu.name))
merged_spe$rank_abund_toad <- NA # add as new column for plotting
merged_spe$rank_abund_toad[order.abund_TOAD] <- 1:nrow(merged_spe)

# calculate rank abundance in Environmental samples of each OTU
order.abund_ENV <- order(merged_spe$rel_abund_env,as.vector(merged_spe$otu.name))
merged_spe$rank_abund_env <- NA # add as new column for plotting
merged_spe$rank_abund_env[order.abund_ENV] <- 1:nrow(merged_spe)


# filter out only the OTU names and relative abundance data
list_OTUs <- list("OTU_15", "OTU_166", "OTU_73", "OTU_1", "OTU_25",
                  "OTU_4", "OTU_58", "OTU_34", "OTU_8", "OTU_99",
                  "OTU_3", "OTU_177", "OTU_92", "OTU_79", "OTU_148",
                  "OTU_34", "OTU_81", "OTU_14", "OTU83", "OTU_166",
                  "OTU_9")
merged_filt_spe <- merged_spe %>%
  dplyr::filter(otu.name %in% list_OTUs)

# plot significant ANCOM data
# rank abundance plot of just the OTU's in toads
rank_toads <- ggplot(merged_filt_spe) + 
  geom_point(aes(x=rank_abund_toad, y=rel_abund_toad, color=otu.name)) +
  xlab("rank abundance on toad samples") +
  ylab("relative abundance on toad samples")

# rank abundance plot of just the OTU's in env
rank_env <- ggplot(merged_filt_spe) + 
  geom_point(aes(x=rank_abund_env, y=rel_abund_env, color=otu.name)) +
  xlab("rank abundance on environmental samples") +
  ylab("relative abundance on environmental samples")

# rank abundance of each of toad and env on same graph
rank_plot <- ggplot(merged_filt_spe) + 
  geom_point(aes(x=rank_abund_env, y=rank_abund_toad, color=otu.name)) +
  xlab("rank abundance on environmental samples") +
  ylab("rank abundance on toad samples") +
  xlim(0,550) + ylim(0,550) +
  geom_abline(intercept=0, slope=1, color="red", linetype="dashed")

ggsave(paste0(work_dir,"Figs_Tables/Symbionts_fig/rank_dotplot.png"), rank_plot, dpi=300,
        width=10, height=7)
```


## Miscellaneous

Here, I'm making a graph showing where certain sample types fell out in the various processing steps.

```{r}
drops <- read.delim(paste0(work_dir,"Figs_Tables/Metadata_dropouts_R.txt"))

#reorder the axes
all_sampletypes <- c("Upland_Soil", "Sediment", "Water", "Egg", "Tadpole", "Metamorph", "Subadult", "Adult", "Mouth", "Tadpole_Dead", "Glove", "Sterile_water", "NoTemplateControl")
drops$Life_Stage_Simplified <- factor(drops$Life_Stage_Simplified, 
                                      levels=all_sampletypes)
madetosteplist <- c("noDNA", "filtSeq", "filtR", "Final")
drops$Made_to_step <- factor(drops$Made_to_step, levels=madetosteplist)

#make graph
drop_graph <- ggplot(drops, aes(fill=Life_Stage_Simplified, 
                                x=Made_to_step)) +
  geom_bar(position="dodge", stat="count") +
  labs(fill="Sample Type", x="Reasons for samples dropping out",
       y="Count") +
  scale_x_discrete(labels=c("PCR did not amplify target gene",
                            "Filtered during sequence processing 
                            (low reads or quality)",
                            "Filtered during R analysis 
                            (some sample types not useful 
                            for answering hypotheses)",
                            "Final dataset")) +
  theme(axis.text.x = element_text(angle=0, hjust=0.5, size=14),
        axis.title = element_text(size=16),
        axis.text.y = element_text(size=14),
        legend.text = element_text(size=14),
        legend.title = element_text(size=16))
drop_graph
# ggsave("drop_graph.jpg", drop_graph, width=21, height=10)
```



